src/com/cdd/bao/template/Vocabulary.java

```diff
@@ -492,28 +493,37 @@ else if (predicate.equals(altLabel1) || predicate.equals(altLabel2) || predicate
 			Statement stmt = iter.next();
 			String uri = stmt.getSubject().getURI(), label = stmt.getObject().asLiteral().getString();
 			uriToLabel.put(uri, label);
-			String[] list = labelToURI.get(label);
-			if (list != null)
-			{
-				list = Arrays.copyOf(list, list.length + 1);
-				list[list.length - 1] = uri;
-				labelToURI.put(label, list);
-			}
-			else labelToURI.put(label, new String[]{uri});
+			HashSet<String> set = labelToURI.get(label);
+			set.add(uri); // adding the uri to the last element of the set
+			labelToURI.put(label, set);
 		}
 
 		// go over the label-to-URI list and whenever there are multiple cases, try to favour the BAO version first (expect a few to
 		// slip through though)
 		for (String label : labelToURI.keySet())
 		{
-			String[] list = labelToURI.get(label);
-			if (list.length == 1) continue;
-			Arrays.sort(list);
+			HashSet<String> set = labelToURI.get(label);
+			if (set.size() == 1) continue;
+			List<String> list = new ArrayList<>(set);
+			Collections.sort(list);
 			int idx = -1;
-			for (int n = 0; n < list.length; n++) if (list[n].startsWith(ModelSchema.PFX_BAO)) {idx = n; break;}
-			// NOTE: if there's more than one BAO-based label, that with the lowest sort order will be retained; there are a 
-			// couple of these in the list
-			if (idx >= 0) labelToURI.put(label, new String[]{list[idx]});
+			for (int n = 0; n < list.size(); n++)
+			{
+				if (list.get(n).startsWith(ModelSchema.PFX_BAO)) 
+				{
+					idx = n;
+					// NOTE: if there's more than one BAO-based label, that with the lowest sort order will be retained; there are a 
+					// couple of these in the list
+					if (idx >= 0) 
+					{
+						HashSet<String> singleLabel = new HashSet<>();
+						singleLabel.add(list.get(idx));
+						labelToURI.put(label, singleLabel);
+					}
+					break; // got the single label at the lowest sort order, so stop looking for others
+				}
+			}
+
```

Comment:
You combine the two statements of the original code into one which is ok. However, this means that you can get rid of some code. Think about what values `idx` can  have; it's always `>= 0` and it's always equal to `n`.
```
for (int n = 0; n < list.size(); n++)
{
  if (list.get(n).startsWith(ModelSchema.PFX_BAO)) 
  {
    // NOTE: if there's more than one BAO-based label, that with the lowest sort order will be retained; 
    // there are a couple of these in the list
    Set<String> singleLabel = new HashSet<>();  // or TreeSet
    singleLabel.add(list.get(idx));
    labelToURI.put(label, singleLabel);
    break; // got the single label at the lowest sort order, so stop looking for others
  }
}
```
Now we see that we don't really need `n`
```
for (String uri : list)
{
  if (uri.startsWith(ModelSchema.PFX_BAO)) 
  {
    // NOTE: if there's more than one BAO-based label, that with the lowest sort order will be retained; 
    // there are a couple of these in the list
    Set<String> singleLabel = new HashSet<>();  // or TreeSet
    singleLabel.add(uri);
    labelToURI.put(label, singleLabel);
    break; // got the single label at the lowest sort order, so stop looking for others
  }
}
```